### ISSUE TITLE
Identity Jacobian Matrix Decorator

### ISSUE THREAD

**Issue by pharringtonp19** [google/jax#8710]
### Proposal 

A decorator that we could apply to a vector-valued function which would tell the AD(reverse/forward) to "implicitly" overwrite the jacobian matrix with the identity matrix. I say "implicitly" because I realize that AD does not form the jacobain matrix but rather vjps, or jvps. 

### Motivation
In bi-level optimization, we often want to differentiate through our parameter updates. It would be great to have the option to replace the jacobians associated with these update functions with the identity matrix. Such an option would enable users to experiment with MAML and First Order MAML with little changes to their code. I believe the current alternative (I might be wrong here!) would be to implement custom jvps/vjps of the update function in optax.

**Comment by clemisch** [comment]
Maybe I don't fully understand your use case, but there is `jax.lax.stop_gradient`:

```
Stops gradient computation.

Operationally ``stop_gradient`` is the identity function, that is, it returns
argument `x` unchanged. However, ``stop_gradient`` prevents the flow of
gradients during forward or reverse-mode automatic differentiation. If there
are multiple nested gradient computations, ``stop_gradient`` stops gradients
for all of them.
```

**Comment by pharringtonp19** [comment]
@clemisch Thanks for the suggestion. I think `stop_gradient` covers a broad set of use cases, but because of the following statement,  it doesn't exactly address my particular use case (but perhaps it's a good starting point to build from!)
>  If there are multiple nested gradient computations, stop_gradient stops gradients for all of them.

Below, I have a toy example that highlights the issue. In short, when differentiating through a composition of functions, I would like to be able to replace an "intermediate" jacobian matrix with the identity matrix. I don't believe `stop_gradient` allows for this. 

```
import jax 
import jax.numpy as jnp 
from jax import custom_vjp 

l = lambda x: jnp.cos(x)
g = lambda x: jnp.sin(x)
f = lambda x: jnp.tanh(x)
h = lambda x: f(g(l(x)))

x = 1.
print(f"Desired Gradient: {jax.grad(f)(g(l(x)))*jax.grad(l)(x)}")
print(f"Standard Jax Gradient: {jax.grad(h)(x)}")

Desired Gradient: -0.6529207229614258
Standard Jax Gradient: -0.559914231300354
```

The current workaround would be to use `custom_vjp` as done below:

```
@custom_vjp 
def g(x):
  return jnp.sin(x)

def f_fwd(x):
  return g(x), (jnp.cos(x),)

def f_bwd(res, g):
  cos_x, = res 
  return (g,)

g.defvjp(f_fwd, f_bwd)

print(f"Standard Jax Gradient With Custom VJP: {jax.grad(h)(x)}")

Standard Jax Gradient With Custom VJP: -0.6529207229614258
```

In my particular use case, `g` is the application of potentially multiple updates of gradient descent. 

Option 1 would then be to construct `custom_vjps` for optax specified update functions.
Option 2 might be to ask the optax folks if they have a workaround/suggestion
Option 3 would be to extend `stop_gradients` to this case where we only want the identity matrix to replace intermediate jacobian matricies.

I don't know if option 3 is feasible, but it would be great if it is!

**Comment by hawkinsp** [comment]
I'd probably use a `custom_jvp` here rather than a `custom_vjp`, since it's a little simpler to define:
```
@jax.custom_jvp 
def g(x):
  return jnp.sin(x)

@g.defjvp
def f_fwd(primals, tangents):
  x, = primals
  x_dot, = tangents
  return g(x), x_dot
```

If you need it in decorator form, perhaps something like this:

```
def identity_jacobian(f):
  @jax.custom_jvp
  @functools.wraps(f)
  def wrapper(*args):
    return f(*args)

  @wrapper.defjvp
  def customjvp(primals, tangents):
    return wrapper(*primals), tangents[0]

  return wrapper


g = identity_jacobian(jnp.sin)
  
print(f"Standard Jax Gradient With Custom VJP: {jax.grad(h)(x)}")
```

Given it seems easy enough to write this via a custom gradient in user code, is there a reason we would need to add this to JAX?

**Comment by hawkinsp** [comment]
Closing since it seems that there's nothing to do on the JAX side. Please feel free to reopen if there's some action we should take.

**Comment by pharringtonp19** [comment]
@hawkinsp Thanks for closing this! 

It appears to be really easy to write custom jvp/vjps for optax based transformations so I don't think there's any need to have a jax specific implementation -- I should have just asked the optax team first! Thanks for the help
